_posts/2024-02-14-provisioning-ubuntu-autoinstall.md

```diff
@@ -0,0 +1,361 @@
+    disable_components: [multiverse]
+    disable_suites: [backports,security,updates]
+  user-data:
+    disable_root: false
+    fqdn: ubuntu2204.example.com
+    users:
+    - name: root
+      gecos: root
+      lock-passwd: false
+      hashed_passwd: _My_Password_
+      ssh_authorized_keys: ["ssh-rsa _My_Foreman_Proxy_SSH_Key_ foreman-proxy@foreman.example.com"]
+  keyboard:
+    layout: us
+    toggle: null
+    variant: ''
+  locale: en_US.UTF-8
+  network:
+    version: 2
+    ethernets:
+      ens160:
+        dhcp4: true
+        dhcp6: false
+  ssh:
+    allow-pw: true
+    install-server: true
+  updates: security
+
+  storage:
+    layout:
+      name: lvm
+  late-commands:
+  - wget -Y off http://foreman.example.com/unattended/finish -O /target/tmp/finish.sh
+  - curtin in-target -- chmod +x /tmp/finish.sh
+  - curtin in-target -- /tmp/finish.sh
+```
+
+## Behind the Scenes
+
+Behind the scenes, there are changes in Ubuntu, Foreman, and the provisioning workflow.
+
+### New Installer
+
+With Ubuntu 20.04.3+, the server installer changed from `debian-installer` to `Subiquity` installer.
+This installer brings various changes:
+
+* UEFI/BIOS boot templates require different Kernel parameters
+* Installation media file layout changed, for example the file structure in the ISO image
+* Unattended installations require a different config file format
+
+This means that Foreman has to search for the boot files in different paths depending on the Ubuntu version and provisioning method to provide the boot files during the host deployment.
+This is technically independent of the used installation mechanism, but the provisioning templates associated with the OS and the installation mechanism of the OS have to match.
+
+Overall, provisioning hosts running Ubuntu 20.04.3+ requires different provisioning templates, a way to handle the ISO image and boot files on Foreman Server and Smart Proxy Servers, and a user data template for Autoinstall.
+
+### Changes in Foreman
+
+- With `Autoinstall`, you need a new type of template called "user data template" similar to cloud deployments. (before: `Preseed` templates)
+- Provide the ISO image and extracted ISO image. (before: installation with `debian-installation` was based on a repository)
+
+With these two changes, the location of the ISO image and the Autoinstall template must be passed to the Kernel.
+Therefore, the new `PXELinux Autoinstall` and `PXEGrub2 Autoinstall` templates contain the URLs where to find the user data template and ISO image.
+
+### Boot Files: Linux Kernel and `initrd`
+
+- For Ubuntu with major OS version `18.04.z` or `20.04` and with minor OS version `0`, `1`, or `2`: Foreman searches for `initrd.gz` and `linux` in `dists/bionic/main/installer-amd64/current/images/netboot/ubuntu-installer/amd64/` for 18.04 and `dists/focal/main/installer-amd64/current/legacy-images/netboot/ubuntu-installer/amd64/` for 20.04.
```

Comment:
```suggestion
- For Ubuntu with major OS version `18.04` or `20.04.z` with minor OS version `0`, `1`, or `2`: Foreman searches for `initrd.gz` and `linux` in `dists/bionic/main/installer-amd64/current/images/netboot/ubuntu-installer/amd64/` for 18.04 and `dists/focal/main/installer-amd64/current/legacy-images/netboot/ubuntu-installer/amd64/` for 20.04.
```

That reads better to me, as you want to say "(any 18.04) or (20.04.z with z < 3)", right?

Reply:
Yes, applied.